go-controller/pkg/cni/cnishim.go

```diff
@@ -0,0 +1,107 @@
+package cni
+
+// contains code for cnishim - one that gets called as the cni Plugin
+// This does not do the real cni work. This is just the client to the cniserver
+// that does the real work.
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+// Plugin is the structure to hold the endpoint information and the corresponding
+// functions to use it
+type Plugin struct {
+	socketPath string
+}
+
+// NewCNIPlugin creates the internal Plugin object
+func NewCNIPlugin(socketPath string) *Plugin {
+	if len(socketPath) == 0 {
+		socketPath = serverSocketPath
+	}
+	return &Plugin{socketPath: socketPath}
+}
+
+// Create and fill a Request with this Plugin's environment and stdin which
+// contain the CNI variables and configuration
+func newCNIRequest(args *skel.CmdArgs) *Request {
+	envMap := make(map[string]string)
+	for _, item := range os.Environ() {
+		idx := strings.Index(item, "=")
+		if idx > 0 {
+			envMap[strings.TrimSpace(item[:idx])] = item[idx+1:]
+		}
+	}
+
+	return &Request{
+		Env:    envMap,
+		Config: args.StdinData,
+	}
+}
+
+// Send a CNI request to the CNI server via JSON + HTTP over a root-owned unix socket,
+// and return the result
+func (p *Plugin) doCNI(url string, req *Request) ([]byte, error) {
+	data, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal CNI request %v: %v", req, err)
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			Dial: func(proto, addr string) (net.Conn, error) {
+				return net.Dial("unix", p.socketPath)
+			},
+		},
+	}
+
+	resp, err := client.Post(url, "application/json", bytes.NewReader(data))
```

Comment:
@dcbw Do we need to do this in hostNs? That is not portable to Windows, so I removed it because cnishim does not switch namespaces. Agree?

Reply:
I meant a containernetworking.HostNS.Do() block.